File: code/ts/modular_api_sqlserver/src/dbClient.ts
export enum DbCommandKind {
  query = 'query',
  execute = 'execute',
  batch = 'batch',
  scalar = 'scalar',
}

export enum DbFailureKind {
  connectivity = 'connectivity',
  timeout = 'timeout',
  authentication = 'authentication',
  authorization = 'authorization',
  constraint = 'constraint',
  conflict = 'conflict',
  notFound = 'not_found',
  serialization = 'serialization',
  cancelled = 'cancelled',
  unknown = 'unknown',
}

export enum DbHealthStatus {
  healthy = 'healthy',
  unhealthy = 'unhealthy',
}

export class DbConnectionSettings {
  public readonly host: string;
  public readonly port: number;
  public readonly database: string;
  public readonly username: string;
  public readonly password: string;
  public readonly driver: string;
  public readonly options: Record<string, unknown>;

  public constructor(options: {
    host: string;
    port: number;
    database: string;
    username: string;
    password: string;
    driver: string;
    options?: Record<string, unknown>;
  }) {
    this.host = options.host;
    this.port = options.port;
    this.database = options.database;
    this.username = options.username;
    this.password = options.password;
    this.driver = options.driver;
    this.options = { ...(options.options ?? {}) };
  }

  public static fromEnvironment(
    environment: Record<string, string | undefined> = process.env,
  ): DbConnectionSettings {
    const parsedPort = Number.parseInt(environment.MODULAR_API_SQLSERVER_PORT ?? '', 10);

    return new DbConnectionSettings({
      host: environment.MODULAR_API_SQLSERVER_HOST ?? '127.0.0.1',
      port: Number.isNaN(parsedPort) ? 14333 : parsedPort,
      database: environment.MODULAR_API_SQLSERVER_DATABASE ?? 'modular_api_graphql_v1',
      username: environment.MODULAR_API_SQLSERVER_USERNAME ?? 'sa',
      password: environment.MODULAR_API_SQLSERVER_PASSWORD ?? 'ModularApi_dev_StrongPass1',
      driver: environment.MODULAR_API_SQLSERVER_DRIVER ?? 'ODBC Driver 17 for SQL Server',
    });
  }

  public get engineId(): string {
    return 'sqlserver';
  }

  public get redactedSummary(): string {
    return `${this.engineId}://${this.username}@${this.host}:${this.port}/${this.database}?driver=${this.driver}`;
  }
}

export class DbCommand {
  public readonly kind: DbCommandKind;
  public readonly text: string;
  public readonly parameters: unknown[];
  public readonly label?: string;

  public constructor(options: {
    kind: DbCommandKind;
    text: string;
    parameters?: unknown[];
    label?: string;
  }) {
    this.kind = options.kind;
    this.text = options.text;
    this.parameters = [...(options.parameters ?? [])];
    this.label = options.label;
  }
}

export class DbExecutionMetadata {
  public readonly duration: number;
  public readonly commandLabel?: string;
  public readonly rowCount?: number;
  public readonly affectedCount?: number;

  public constructor(options: {
    duration: number;
    commandLabel?: string;
    rowCount?: number;
    affectedCount?: number;
  }) {
    this.duration = options.duration;
    this.commandLabel = options.commandLabel;
    this.rowCount = options.rowCount;
    this.affectedCount = options.affectedCount;
  }
}

export class DbRowSet {
  public readonly rows: Array<Record<string, unknown>>;
  public readonly metadata: DbExecutionMetadata;

  public constructor(options: { rows: Array<Record<string, unknown>>; metadata: DbExecutionMetadata }) {
    this.rows = [...options.rows];
    this.metadata = options.metadata;
  }
}

export class DbExecutionSummary {
  public readonly affectedCount: number;
  public readonly metadata: DbExecutionMetadata;

  public constructor(options: { affectedCount: number; metadata: DbExecutionMetadata }) {
    this.affectedCount = options.affectedCount;
    this.metadata = options.metadata;
  }
}

export class DbScalar<T> {
  public readonly value: T;
  public readonly metadata: DbExecutionMetadata;

  public constructor(options: { value: T; metadata: DbExecutionMetadata }) {
    this.value = options.value;
    this.metadata = options.metadata;
  }
}

export class DbFailure {
  public readonly kind: DbFailureKind;
  public readonly code: string;
  public readonly message: string;
  public readonly retryable: boolean;
  public readonly transient: boolean;
  public readonly details?: unknown;
  public readonly causeSummary?: string;

  public constructor(options: {
    kind: DbFailureKind;
    code: string;
    message: string;
    retryable: boolean;
    transient: boolean;
    details?: unknown;
    causeSummary?: string;
  }) {
    this.kind = options.kind;
    this.code = options.code;
    this.message = options.message;
    this.retryable = options.retryable;
    this.transient = options.transient;
    this.details = options.details;
    this.causeSummary = options.causeSummary;
  }
}

export class DbResult<T> {
  private readonly innerValue?: T;
  private readonly innerFailure?: DbFailure;

  private constructor(value?: T, failure?: DbFailure) {
    this.innerValue = value;
    this.innerFailure = failure;
  }

  public static success<T>(value: T): DbResult<T> {
    return new DbResult<T>(value, undefined);
  }

  public static failure<T>(failure: DbFailure): DbResult<T> {
    return new DbResult<T>(undefined, failure);
  }

  public get isSuccess(): boolean {
    return this.innerFailure === undefined;
  }

  public get isFailure(): boolean {
    return this.innerFailure !== undefined;
  }

  public get value(): T {
    if (this.innerFailure !== undefined) {
      throw new Error('DbResult does not contain a success value.');
    }
    return this.innerValue as T;
  }

  public get failure(): DbFailure {
    if (this.innerFailure === undefined) {
      throw new Error('DbResult does not contain a failure value.');
    }
    return this.innerFailure;
  }

  public map<R>(transform: (value: T) => R): DbResult<R> {
    if (this.isFailure) {
      return DbResult.failure(this.failure);
    }
    return DbResult.success(transform(this.value));
  }

  public flatMap<R>(transform: (value: T) => DbResult<R>): DbResult<R> {
    if (this.isFailure) {
      return DbResult.failure(this.failure);
    }
    return transform(this.value);
  }

  public mapFailure(transform: (failure: DbFailure) => DbFailure): DbResult<T> {
    if (this.isSuccess) {
      return DbResult.success(this.value);
    }
    return DbResult.failure(transform(this.failure));
  }

  public getOrThrow(message?: string): T {
    if (this.isFailure) {
      throw new Error(message ?? this.failure.message);
    }
    return this.value;
  }
}

export class DbProviderDescription {
  public readonly engineId: string;
  public readonly database: string;
  public readonly redactedSummary: string;
  public readonly ownsResources: boolean;

  public constructor(options: {
    engineId: string;
    database: string;
    redactedSummary: string;
    ownsResources: boolean;
  }) {
    this.engineId = options.engineId;
    this.database = options.database;
    this.redactedSummary = options.redactedSummary;
    this.ownsResources = options.ownsResources;
  }
}

export class DbSessionLease<S> {
  public readonly session: S;
  public readonly ownedByPackage: boolean;
  private readonly releaser: () => Promise<DbResult<void>>;

  public constructor(options: {
    session: S;
    ownedByPackage: boolean;
    releaser: () => Promise<DbResult<void>>;
  }) {
    this.session = options.session;
    this.ownedByPackage = options.ownedByPackage;
    this.releaser = options.releaser;
  }

  public async release(): Promise<DbResult<void>> {
    if (!this.ownedByPackage) {
      return DbResult.success<void>(undefined);
    }
    return this.releaser();
  }
}

export interface DbSessionProvider<S> {
  acquire(): Promise<DbResult<DbSessionLease<S>>>;
  close(): Promise<DbResult<void>>;
  describe(): DbProviderDescription;
}

export interface DbCommandExecutor<S> {
  query(session: S, command: DbCommand): Promise<DbResult<DbRowSet>>;
  execute(session: S, command: DbCommand): Promise<DbResult<DbExecutionSummary>>;
  scalar<T>(session: S, command: DbCommand): Promise<DbResult<DbScalar<T>>>;
}

export class DbTransactionContext<S> {
  public readonly settings: DbConnectionSettings;
  public readonly session: S;
  public readonly commandExecutor: DbCommandExecutor<S>;

  public constructor(options: {
    settings: DbConnectionSettings;
    session: S;
    commandExecutor: DbCommandExecutor<S>;
  }) {
    this.settings = options.settings;
    this.session = options.session;
    this.commandExecutor = options.commandExecutor;
  }

  public query(command: DbCommand): Promise<DbResult<DbRowSet>> {
    return this.commandExecutor.query(this.session, command);
  }

  public execute(command: DbCommand): Promise<DbResult<DbExecutionSummary>> {
    return this.commandExecutor.execute(this.session, command);
  }

  public scalar<T>(command: DbCommand): Promise<DbResult<DbScalar<T>>> {
    return this.commandExecutor.scalar<T>(this.session, command);
  }
}

export interface DbTransactionRunner<S> {
  run<T>(
    context: DbTransactionContext<S>,
    body: (context: DbTransactionContext<S>) => Promise<DbResult<T>>,
  ): Promise<DbResult<T>>;
}

export class DbRepositoryContext<S> {
  public readonly settings: DbConnectionSettings;
  public readonly sessionProvider: DbSessionProvider<S>;
  public readonly commandExecutor: DbCommandExecutor<S>;
  public readonly transactionRunner: DbTransactionRunner<S>;

  public constructor(options: {
    settings: DbConnectionSettings;
    sessionProvider: DbSessionProvider<S>;
    commandExecutor: DbCommandExecutor<S>;
    transactionRunner: DbTransactionRunner<S>;
  }) {
    this.settings = options.settings;
    this.sessionProvider = options.sessionProvider;
    this.commandExecutor = options.commandExecutor;
    this.transactionRunner = options.transactionRunner;
  }
}

export abstract class DbRepository<S> {
  public constructor(public readonly context: DbRepositoryContext<S>) {}

  public query(command: DbCommand): Promise<DbResult<DbRowSet>> {
    return withLease(this.context.sessionProvider, (lease) =>
      this.context.commandExecutor.query(lease.session, command),
    );
  }

  public execute(command: DbCommand): Promise<DbResult<DbExecutionSummary>> {
    return withLease(this.context.sessionProvider, (lease) =>
      this.context.commandExecutor.execute(lease.session, command),
    );
  }

  public scalar<T>(command: DbCommand): Promise<DbResult<DbScalar<T>>> {
    return withLease(this.context.sessionProvider, (lease) =>
      this.context.commandExecutor.scalar<T>(lease.session, command),
    );
  }

  public transaction<T>(
    body: (context: DbTransactionContext<S>) => Promise<DbResult<T>>,
  ): Promise<DbResult<T>> {
    const client = new DbClient({
      settings: this.context.settings,
      sessionProvider: this.context.sessionProvider,
      commandExecutor: this.context.commandExecutor,
      transactionRunner: this.context.transactionRunner,
    });
    return client.transaction(body);
  }
}

export class DbClient<S> {
  public readonly settings: DbConnectionSettings;
  public readonly sessionProvider: DbSessionProvider<S>;
  public readonly commandExecutor: DbCommandExecutor<S>;
  public readonly transactionRunner: DbTransactionRunner<S>;

  public constructor(options: {
    settings: DbConnectionSettings;
    sessionProvider: DbSessionProvider<S>;
    commandExecutor: DbCommandExecutor<S>;
    transactionRunner: DbTransactionRunner<S>;
  }) {
    this.settings = options.settings;
    this.sessionProvider = options.sessionProvider;
    this.commandExecutor = options.commandExecutor;
    this.transactionRunner = options.transactionRunner;
  }

  public query(command: DbCommand): Promise<DbResult<DbRowSet>> {
    return withLease(this.sessionProvider, (lease) =>
      this.commandExecutor.query(lease.session, command),
    );
  }

  public execute(command: DbCommand): Promise<DbResult<DbExecutionSummary>> {
    return withLease(this.sessionProvider, (lease) =>
      this.commandExecutor.execute(lease.session, command),
    );
  }

  public scalar<T>(command: DbCommand): Promise<DbResult<DbScalar<T>>> {
    return withLease(this.sessionProvider, (lease) =>
      this.commandExecutor.scalar<T>(lease.session, command),
    );
  }

  public async transaction<T>(
    body: (context: DbTransactionContext<S>) => Promise<DbResult<T>>,
  ): Promise<DbResult<T>> {
    const leaseResult = await this.sessionProvider.acquire();
    if (leaseResult.isFailure) {
      return DbResult.failure(leaseResult.failure);
    }

    const lease = leaseResult.value;
    const context = new DbTransactionContext({
      settings: this.settings,
      session: lease.session,
      commandExecutor: this.commandExecutor,
    });
    const result = await this.transactionRunner.run(context, body);
    const releaseResult = await lease.release();

    if (result.isFailure) {
      return DbResult.failure(result.failure);
    }
    if (releaseResult.isFailure) {
      return DbResult.failure(releaseResult.failure);
    }

    return result;
  }

  public repositoryContext(): DbRepositoryContext<S> {
    return new DbRepositoryContext({
      settings: this.settings,
      sessionProvider: this.sessionProvider,
      commandExecutor: this.commandExecutor,
      transactionRunner: this.transactionRunner,
    });
  }

  public close(): Promise<DbResult<void>> {
    return this.sessionProvider.close();
  }

  public describe(): DbProviderDescription {
    return this.sessionProvider.describe();
  }
}

export class DbHealthReport {
  public readonly status: DbHealthStatus;
  public readonly responseTime: number;
  public readonly redactedSummary: string;
  public readonly details?: string;

  public constructor(options: {
    status: DbHealthStatus;
    responseTime: number;
    redactedSummary: string;
    details?: string;
  }) {
    this.status = options.status;
    this.responseTime = options.responseTime;
    this.redactedSummary = options.redactedSummary;
    this.details = options.details;
  }
}

export class DbHealthContributor<S> {
  public readonly client: DbClient<S>;
  public readonly probeCommand: DbCommand;

  public constructor(options: {
    client: DbClient<S>;
    probeCommand?: DbCommand;
  }) {
    this.client = options.client;
    this.probeCommand =
      options.probeCommand ??
      new DbCommand({
        kind: DbCommandKind.scalar,
        text: 'SELECT 1',
        label: 'db.health',
      });
  }

  public async probe(): Promise<DbHealthReport> {
    const startedAt = Date.now();
    const result = await this.client.scalar(this.probeCommand);
    const responseTime = Date.now() - startedAt;

    if (result.isSuccess) {
      return new DbHealthReport({
        status: DbHealthStatus.healthy,
        responseTime,
        redactedSummary: this.client.describe().redactedSummary,
      });
    }

    return new DbHealthReport({
      status: DbHealthStatus.unhealthy,
      responseTime,
      redactedSummary: this.client.describe().redactedSummary,
      details: result.failure.code,
    });
  }
}

export class DbGraphqlSupport<S> {
  public readonly catalogProvider: unknown;
  public readonly readExecutor: unknown;
  public readonly healthContributor: DbHealthContributor<S>;
  public readonly sourceDigestFactory?: unknown;
  public readonly artifactLoader?: unknown;
  public readonly capabilityRegistration?: unknown;

  public constructor(options: {
    catalogProvider: unknown;
    readExecutor: unknown;
    healthContributor: DbHealthContributor<S>;
    sourceDigestFactory?: unknown;
    artifactLoader?: unknown;
    capabilityRegistration?: unknown;
  }) {
    this.catalogProvider = options.catalogProvider;
    this.readExecutor = options.readExecutor;
    this.healthContributor = options.healthContributor;
    this.sourceDigestFactory = options.sourceDigestFactory;
    this.artifactLoader = options.artifactLoader;
    this.capabilityRegistration = options.capabilityRegistration;
  }
}

async function withLease<S, T>(
  sessionProvider: DbSessionProvider<S>,
  operation: (lease: DbSessionLease<S>) => Promise<DbResult<T>>,
): Promise<DbResult<T>> {
  const leaseResult = await sessionProvider.acquire();
  if (leaseResult.isFailure) {
    return DbResult.failure(leaseResult.failure);
  }

  const lease = leaseResult.value;
  const operationResult = await operation(lease);
  const releaseResult = await lease.release();

  if (operationResult.isFailure) {
    return DbResult.failure(operationResult.failure);
  }
  if (releaseResult.isFailure) {
    return DbResult.failure(releaseResult.failure);
  }

  return operationResult;
}

File: code/py/modular_api_sqlserver/src/modular_api_sqlserver/db_client.py
from __future__ import annotations

import os
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Generic, Mapping, Protocol, TypeVar, cast

S = TypeVar("S")
T = TypeVar("T")
R = TypeVar("R")
_MISSING = object()


class DbCommandKind(str, Enum):
    QUERY = "query"
    EXECUTE = "execute"
    BATCH = "batch"
    SCALAR = "scalar"
    PROCEDURE = "procedure"


class DbParameterDirection(str, Enum):
    INPUT = "input"
    OUTPUT = "output"
    INPUT_OUTPUT = "inputOutput"


class DbFailureKind(str, Enum):
    CONNECTIVITY = "connectivity"
    TIMEOUT = "timeout"
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    CONSTRAINT = "constraint"
    CONFLICT = "conflict"
    NOT_FOUND = "not_found"
    SERIALIZATION = "serialization"
    CANCELLED = "cancelled"
    UNKNOWN = "unknown"


class DbHealthStatus(str, Enum):
    HEALTHY = "healthy"
    UNHEALTHY = "unhealthy"


@dataclass(frozen=True, slots=True)
class DbConnectionSettings:
    host: str
    port: int
    database: str
    username: str
    password: str
    driver: str
    options: Mapping[str, object] = field(default_factory=dict)

    @classmethod
    def from_environment(
        cls,
        environment: Mapping[str, str] | None = None,
    ) -> DbConnectionSettings:
        values = os.environ if environment is None else environment
        raw_port = values.get("MODULAR_API_SQLSERVER_PORT", "")

        try:
            port = int(raw_port)
        except ValueError:
            port = 14333

        return cls(
            host=values.get("MODULAR_API_SQLSERVER_HOST", "127.0.0.1"),
            port=port,
            database=values.get("MODULAR_API_SQLSERVER_DATABASE", "modular_api_graphql_v1"),
            username=values.get("MODULAR_API_SQLSERVER_USERNAME", "sa"),
            password=values.get("MODULAR_API_SQLSERVER_PASSWORD", "ModularApi_dev_StrongPass1"),
            driver=values.get(
                "MODULAR_API_SQLSERVER_DRIVER",
                "ODBC Driver 17 for SQL Server",
            ),
        )

    @property
    def engine_id(self) -> str:
        return "sqlserver"

    @property
    def redacted_summary(self) -> str:
        return (
            f"{self.engine_id}://{self.username}@{self.host}:{self.port}/"
            f"{self.database}?driver={self.driver}"
        )


@dataclass(frozen=True, slots=True)
class DbParameter:
    name: str
    value: object | None = None
    direction: DbParameterDirection = DbParameterDirection.INPUT
    type_hint: str | None = None

    @classmethod
    def input(cls, name: str, value: object | None, type_hint: str | None = None) -> DbParameter:
        return cls(name=name, value=value, direction=DbParameterDirection.INPUT, type_hint=type_hint)

    @classmethod
    def output(cls, name: str, type_hint: str | None = None) -> DbParameter:
        return cls(name=name, direction=DbParameterDirection.OUTPUT, type_hint=type_hint)

    @classmethod
    def input_output(
        cls, name: str, value: object | None, type_hint: str | None = None
    ) -> DbParameter:
        return cls(
            name=name,
            value=value,
            direction=DbParameterDirection.INPUT_OUTPUT,
            type_hint=type_hint,
        )


@dataclass(frozen=True, slots=True)
class DbProcedureOutcome:
    return_value: object | None = None
    output_parameters: Mapping[str, object] | None = None


@dataclass(frozen=True, slots=True)
class DbCommand:
    kind: DbCommandKind
    text: str
    parameters: tuple[object, ...] = field(default_factory=tuple)
    label: str | None = None


@dataclass(frozen=True, slots=True)
class DbExecutionMetadata:
    duration: int
    command_label: str | None = None
    row_count: int | None = None
    affected_count: int | None = None


@dataclass(frozen=True, slots=True)
class DbRowSet:
    rows: list[Mapping[str, object]]
    metadata: DbExecutionMetadata
    procedure: DbProcedureOutcome | None = None


@dataclass(frozen=True, slots=True)
class DbExecutionSummary:
    affected_count: int
    metadata: DbExecutionMetadata
    procedure: DbProcedureOutcome | None = None


@dataclass(frozen=True, slots=True)
class DbScalar(Generic[T]):
    value: T
    metadata: DbExecutionMetadata


@dataclass(frozen=True, slots=True)
class DbFailure:
    kind: DbFailureKind
    code: str
    message: str
    retryable: bool
    transient: bool
    details: object | None = None
    cause_summary: str | None = None


class DbResult(Generic[T]):
    def __init__(self, value: object = _MISSING, failure: DbFailure | None = None) -> None:
        self._value = value
        self._failure = failure

    @classmethod
    def success(cls, value: T) -> DbResult[T]:
        return cls(value=value)

    @classmethod
    def from_failure(cls, failure: DbFailure) -> DbResult[T]:
        return cls(failure=failure)

    @property
    def is_success(self) -> bool:
        return self._failure is None

    @property
    def is_failure(self) -> bool:
        return self._failure is not None

    @property
    def value(self) -> T:
        if self._failure is not None or self._value is _MISSING:
            raise RuntimeError("DbResult does not contain a success value.")
        return cast(T, self._value)

    @property
    def failure(self) -> DbFailure:
        if self._failure is None:
            raise RuntimeError("DbResult does not contain a failure value.")
        return self._failure

    def map(self, transform: Callable[[T], R]) -> DbResult[R]:
        if self.is_failure:
            return DbResult.from_failure(self.failure)
        return DbResult.success(transform(self.value))

    def flat_map(self, transform: Callable[[T], DbResult[R]]) -> DbResult[R]:
        if self.is_failure:
            return DbResult.from_failure(self.failure)
        return transform(self.value)

    def map_failure(self, transform: Callable[[DbFailure], DbFailure]) -> DbResult[T]:
        if self.is_success:
            return DbResult.success(self.value)
        return DbResult.from_failure(transform(self.failure))

    def get_or_throw(self, message: str | None = None) -> T:
        if self.is_failure:
            raise RuntimeError(message or self.failure.message)
        return self.value


@dataclass(frozen=True, slots=True)
class DbProviderDescription:
    engine_id: str
    database: str
    redacted_summary: str
    owns_resources: bool


class DbSessionLease(Generic[S]):
    def __init__(
        self,
        *,
        session: S,
        owned_by_package: bool,
        releaser: Callable[[], DbResult[None]],
    ) -> None:
        self.session = session
        self.owned_by_package = owned_by_package
        self._releaser = releaser

    def release(self) -> DbResult[None]:
        if not self.owned_by_package:
            return DbResult.success(None)
        return self._releaser()


class DbSessionProvider(Protocol[S]):
    def acquire(self) -> DbResult[DbSessionLease[S]]: ...

    def close(self) -> DbResult[None]: ...

    def describe(self) -> DbProviderDescription: ...


class DbCommandExecutor(Protocol[S]):
    def query(self, session: S, command: DbCommand) -> DbResult[DbRowSet]: ...

    def execute(self, session: S, command: DbCommand) -> DbResult[DbExecutionSummary]: ...

    def scalar(self, session: S, command: DbCommand) -> DbResult[DbScalar[object]]: ...


@dataclass(frozen=True, slots=True)
class DbTransactionContext(Generic[S]):
    settings: DbConnectionSettings
    session: S
    command_executor: DbCommandExecutor[S]

    def query(self, command: DbCommand) -> DbResult[DbRowSet]:
        return self.command_executor.query(self.session, command)

    def execute(self, command: DbCommand) -> DbResult[DbExecutionSummary]:
        return self.command_executor.execute(self.session, command)

    def scalar(self, command: DbCommand) -> DbResult[DbScalar[T]]:
        return cast(DbResult[DbScalar[T]], self.command_executor.scalar(self.session, command))


class DbTransactionRunner(Protocol[S]):
    def run(
        self,
        context: DbTransactionContext[S],
        body: Callable[[DbTransactionContext[S]], DbResult[T]],
    ) -> DbResult[T]: ...


@dataclass(frozen=True, slots=True)
class DbRepositoryContext(Generic[S]):
    settings: DbConnectionSettings
    session_provider: DbSessionProvider[S]
    command_executor: DbCommandExecutor[S]
    transaction_runner: DbTransactionRunner[S]


class DbRepository(Generic[S]):
    def __init__(self, context: DbRepositoryContext[S]) -> None:
        self.context = context

    def query(self, command: DbCommand) -> DbResult[DbRowSet]:
        return _with_lease(
            self.context.session_provider,
            lambda lease: self.context.command_executor.query(lease.session, command),
        )

    def execute(self, command: DbCommand) -> DbResult[DbExecutionSummary]:
        return _with_lease(
            self.context.session_provider,
            lambda lease: self.context.command_executor.execute(lease.session, command),
        )

    def scalar(self, command: DbCommand) -> DbResult[DbScalar[T]]:
        return cast(
            DbResult[DbScalar[T]],
            _with_lease(
                self.context.session_provider,
                lambda lease: self.context.command_executor.scalar(lease.session, command),
            ),
        )

    def transaction(self, body: Callable[[DbTransactionContext[S]], DbResult[T]]) -> DbResult[T]:
        client = DbClient(
            settings=self.context.settings,
            session_provider=self.context.session_provider,
            command_executor=self.context.command_executor,
            transaction_runner=self.context.transaction_runner,
        )
        return client.transaction(body)


class DbClient(Generic[S]):
    def __init__(
        self,
        *,
        settings: DbConnectionSettings,
        session_provider: DbSessionProvider[S],
        command_executor: DbCommandExecutor[S],
        transaction_runner: DbTransactionRunner[S],
    ) -> None:
        self.settings = settings
        self.session_provider = session_provider
        self.command_executor = command_executor
        self.transaction_runner = transaction_runner

    def query(self, command: DbCommand) -> DbResult[DbRowSet]:
        return _with_lease(
            self.session_provider,
            lambda lease: self.command_executor.query(lease.session, command),
        )

    def execute(self, command: DbCommand) -> DbResult[DbExecutionSummary]:
        return _with_lease(
            self.session_provider,
            lambda lease: self.command_executor.execute(lease.session, command),
        )

    def scalar(self, command: DbCommand) -> DbResult[DbScalar[T]]:
        return cast(
            DbResult[DbScalar[T]],
            _with_lease(
                self.session_provider,
                lambda lease: self.command_executor.scalar(lease.session, command),
            ),
        )

    def transaction(self, body: Callable[[DbTransactionContext[S]], DbResult[T]]) -> DbResult[T]:
        lease_result = self.session_provider.acquire()
        if lease_result.is_failure:
            return DbResult.from_failure(lease_result.failure)

        lease = lease_result.value
        context = DbTransactionContext(
            settings=self.settings,
            session=lease.session,
            command_executor=self.command_executor,
        )
        result = self.transaction_runner.run(context, body)
        release_result = lease.release()

        if result.is_failure:
            return DbResult.from_failure(result.failure)
        if release_result.is_failure:
            return DbResult.from_failure(release_result.failure)
        return result

    def repository_context(self) -> DbRepositoryContext[S]:
        return DbRepositoryContext(
            settings=self.settings,
            session_provider=self.session_provider,
            command_executor=self.command_executor,
            transaction_runner=self.transaction_runner,
        )

    def close(self) -> DbResult[None]:
        return self.session_provider.close()

    def describe(self) -> DbProviderDescription:
        return self.session_provider.describe()


@dataclass(frozen=True, slots=True)
class DbHealthReport:
    status: DbHealthStatus
    response_time: int
    redacted_summary: str
    details: str | None = None


class DbHealthContributor(Generic[S]):
    def __init__(self, *, client: DbClient[S], probe_command: DbCommand | None = None) -> None:
        self.client = client
        self.probe_command = probe_command or DbCommand(
            kind=DbCommandKind.SCALAR,
            text="SELECT 1",
            label="db.health",
        )

    def probe(self) -> DbHealthReport:
        started_at = time.monotonic()
        result = self.client.scalar(self.probe_command)
        response_time = int((time.monotonic() - started_at) * 1000)

        if result.is_success:
            return DbHealthReport(
                status=DbHealthStatus.HEALTHY,
                response_time=response_time,
                redacted_summary=self.client.describe().redacted_summary,
            )

        return DbHealthReport(
            status=DbHealthStatus.UNHEALTHY,
            response_time=response_time,
            redacted_summary=self.client.describe().redacted_summary,
            details=result.failure.code,
        )


@dataclass(frozen=True, slots=True)
class DbGraphqlSupport(Generic[S]):
    catalog_provider: object
    read_executor: object
    health_contributor: DbHealthContributor[S]
    source_digest_factory: object | None = None
    artifact_loader: object | None = None
    capability_registration: object | None = None


def _with_lease(
    session_provider: DbSessionProvider[S],
    operation: Callable[[DbSessionLease[S]], DbResult[T]],
) -> DbResult[T]:
    lease_result = session_provider.acquire()
    if lease_result.is_failure:
        return DbResult.from_failure(lease_result.failure)

    lease = lease_result.value
    operation_result = operation(lease)
    release_result = lease.release()

    if operation_result.is_failure:
        return DbResult.from_failure(operation_result.failure)
    if release_result.is_failure:
        return DbResult.from_failure(release_result.failure)
    return operation_result